src/org/jetbrains/java/decompiler/modules/decompiler/IfPatternMatchProcessor.java

```diff
@@ -47,86 +54,121 @@ private static boolean matchInstanceofRec(Statement statement, RootStatement roo
-                      // TODO: cache this
-                      findVarsInPredecessors(vvs, statement.getIfstat());
-
-                      VarVersionPair var = ((VarExprent) left).getVarVersionPair();
-
-                      // Stop processing if this variable has already been seen
-                      for (VarVersionPair vv : vvs) {
-                        if (var.var == vv.var) {
-                          continue loop;
-                        }
-                      }
-
-                      // Add the exprent to the instanceof exprent and remove it from the inside of the if statement
-                      iof.getLstOperands().add(2, left);
-                      head.getExprents().remove(0);
-
-                      statement.setPatternMatched(true);
-
-                      updated = true;
-                    }
-                  }
-                }
-              }
-            }
-          }
+    if (lastIfTrue != null) {
+      updated = checkBranch(lastIfTrue, statement, statement.getIfEdge().getDestination());
+    }
+
+    if (!updated && lastIfFalse != null) {
+      if (statement.getElseEdge() != null) {
+        updated |= checkBranch(lastIfFalse, statement, statement.getElseEdge().getDestination());
+      } else {
+        var allSuc = statement.getAllSuccessorEdges();
+        if (allSuc.size() == 1) {
+          // In theory, the if branch can 'fall through' to here, but then this branch has multiple predecessors
+          // and will get left alone anyway
+          updated |= checkBranch(lastIfFalse, statement, allSuc.get(0).getDestination());
         }
       }
     }
 
+    if (updated) {
+      // It's possible that the if statement is now empty
+      // if(){;}else{...} -> if(!){...}
+      statement.checkInversion();
+    }
+
     return updated;
   }
 
+  private static boolean checkBranch(Exprent exprent, IfStatement statement, Statement branch) {
+    if (!(exprent instanceof FunctionExprent) || branch.getAllPredecessorEdges().size() != 1) {
+      // We can only inline into 'instanceof', and only if the target branch doesn't have multiple predecessors
+      // TODO: make checking for multiple predecessors less expensive
+      return false;
+    }
+
+    FunctionExprent iof = (FunctionExprent) exprent;
+
+    // Check for instanceof and isn't a pattern match yet
+    if (iof.getFuncType() != FunctionType.INSTANCEOF || iof.getLstOperands().size() != 2) {
+      return false;
+    }
+
+    Exprent source = iof.getLstOperands().get(0);
+    Exprent target = iof.getLstOperands().get(1);
+
+    // Check to make sure there is more than 1 exprent.
+    // More often than not, when there's less than 1 it means it's assigning into a previous value.
+    // TODO: this isn't always the case, handle it properly
+    Statement head = branch.getBasichead();
+
+    if (head == null || head.getExprents() == null) {
```

Comment:
`basichead` for an if statement should never be false, so the first branch isn't needed. We should add an assert to getBasicHead validating this invariant.